README
What is bbdock ?

  bbdock displays one or more PNG icons in the BlackBox/FluxBox slit from
  which you can launch the corresponding applications. It is pretty similar
  in function and appearance to bbbutton which you might want to try if you
  are reading this on a display with a color-depth below 16 bits.

How to compile

  Since you are reading this I assume you already unpacked the archive into
  some directory.

  ./configure
  make
  su
  make install

  For more options try ./configure --help

  You need X11 libraries & headers and libpng-devel. If anything fails,
  just drop me a line at mf@markusfisch.de ;-)

Configuration

  You may invoke bbdock either with your icons specified on command line
  or simply create a short configuration file (.bbdockrc) in your home
  directory.

  The syntax for a command-line-argument or a line in the configuration
  file is equal:

  IMAGEFILE:COMMAND[:WINDOWTITLE;[:IDLE]]

  IMAGEFILE   - should be path and filename of some PNG icon
  COMMAND     - a script or binary to execute
  WINDOWTITLE - is a semicolon-seperated list of case-insensitive
                window-titles of corresponding application-windows. Those
                strings may contain wildcard characters (* and/or ?) to
                exclusively identify a window. By providing this list you
                make the icon exclusive to one instance of course. Clicking
                on already launched icons will raise the corresponding
                window instead of invoking a new instance.
  IDLE        - idle time after triggering one icon in miliseconds, instead
                of using numbers you may also use the terms "lame" (~ 10 s),
                "slow" (~ 5 s), "fast" (~ 500 ms) or "fastest" (~ 250 ms)

  For example, run it from command line this way:

  $ bbdock "~/.icons/firefox.png:firefox:*Firefox"

  Sample out of a ~/.bbdockrc:

  ~/.icons/terminal.png:xterm::1
  ~/.icons/firefox.png:firefox:*Firefox

  To start by configuration file, just run bbdock. Type "bbdock -h" for
  a detailed view of general options.

Requirement

  The images should be not much greater than 48x48 pixels and must be in
  PNG format. Use images with transparency to get the most out of it.

  You need libpng-devel to compile this software. The raise-window
  function is only available to window managers which implement the EWMH
  specification.

Bugs

  Currently there are only two known bugs you might run into:

  If you're using BlackBox > 0.70 you need to patch it in order to make it
  work correctly. Just insert the following code after line 250 into
  blackbox-0.70.0/src/Slit.cc and recompile BlackBox:

      if ((texture.texture() & bt::Texture::Gradient) && frame.pixmap)
        XSetWindowBackgroundPixmap(display, frame.window, frame.pixmap);
      else if ((texture.texture() & bt::Texture::Solid))
        XSetWindowBackground(display, frame.window,
          texture.color1().pixel(screen->screenNumber()));

  FluxBox < 0.9 may fail to raise applications that are exclusive. Update
  to FluxBox 0.9 to fix this problem.

Links

  bbdock - http://bbdock.nethence.com
  BlackBox - http://blackboxwm.sourceforge.net/
  FluxBox - http://www.fluxbox.org/
  BBButton - http://www.angelfire.com/theforce/button/
  wmctrl - http://sweb.cz/tripie/utils/wmctrl/

Licencing

  See the COPYING file

Authors

  See the AUTHORS file


NEWS
bbdock will now compile on modern systems


src/Dock.cc
/*
 * Copyright (c) 2006 Markus Fisch
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. Neither the name of Markus Fisch nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS ``AS IS''
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */
#include <sys/time.h>
#include <sys/types.h>
#include <stdlib.h>
#include <stdio.h>
#include <unistd.h>
#include <iostream>
#include <string>

#include "Dock.hh"
#include "Slot.hh"
#include "WindowManager.hh"

using namespace std;
using namespace bbdock;

const char *Dock::ChangeIconMessage = "_BBDOCK_CHANGE_ICON_";
const char *Dock::ExecuteIconMessage = "_BBDOCK_EXECUTE_ICON_";

/**
 * Initialize dock
 *
 * @param s - settings
 */
Dock::Dock( Dock::Settings &s ) :
	settings( s ),
	first( 0 )
{
	if(	!(display = XOpenDisplay( 0 )) )
		throw "Unable to open display";

	screen = DefaultScreen( display );
	screenwidth = DisplayWidth( display, screen );
	screenheight = DisplayHeight( display, screen );

	gc = XDefaultGC( display, screen );

	// determine bits per pixel of an screen-compatible XImage, does not 
	// need to be the same as the color-depth of the screen
	{
		XPixmapFormatValues *pf;
		int formats;

		bitsperpixel = DefaultDepth( display, screen );
		pf = XListPixmapFormats( display, &formats );

		if( formats )
		{
			int format;

			for( format = formats;
			     format--; )
				if( pf[format].depth == bitsperpixel )
				{
					bitsperpixel = pf[format].bits_per_pixel;
					break; 
				}

			XFree( (char *) pf );
		}
	}
}

/**
 * Free resources  
 */
Dock::~Dock()
{
	XCloseDisplay( display );
}

/**
 * Add an icon to the dock
 *
 * @param icon - Icon object
 */
bool Dock::add( Icon *icon )
{
	if( !icon )
		return false;

	first = new Slot( *this, *icon, first );

	return true;
}

/**
 * Run the dock
 */
void Dock::run()
{
	if( !first )
		return;

	for( int xfd = ConnectionNumber( display ); ; )
	{
		if( !XPending( display ) )
		{
			Slot *slot;

			for( slot = first;
				slot;
				slot = slot->getNext() )
				slot->tick();

			fd_set rfds;
			struct timeval tv;

			FD_ZERO( &rfds );
			FD_SET( xfd, &rfds );

			tv.tv_sec = 0;
			tv.tv_usec = Suspend;

			select( xfd+1, &rfds, 0, 0, &tv );
			continue;
		}

		XEvent event;

		bzero( &event, sizeof( event ) );
		XNextEvent( display, &event );

		Slot *slot;

		for( slot = first;
			slot;
			slot = slot->getNext() )
			if( event.xany.window == slot->getWindow() )
				break;

		if( !slot )
		{
			if( event.type == ClientMessage )
				client( &event );

			continue;
		}

		switch( event.type )
		{
			case Expose:
				if( !event.xexpose.count )
					slot->draw();
				break;
			case VisibilityNotify:
				slot->setVisibility( event.xvisibility.state );
				break;
			case ConfigureNotify:
				slot->invalidate();
				slot->draw( true );
				break;
			case SelectionNotify:
				if( event.xselection.property != None )
				{
					Atom type;
					int format;
					unsigned long len;
					unsigned long after;
					unsigned long bytes;
					unsigned char *data;

					if( XGetWindowProperty( display, event.xany.window, 
						event.xselection.property, 0, 0, 0, AnyPropertyType, 
						&type, &format, &len, &bytes, &data ) != Success ||
						!bytes )
						break;

					if( XGetWindowProperty( display, event.xany.window, 
						event.xselection.property, 0, bytes, 0, 
						AnyPropertyType, &type, &format, &len, &after, 
						&data ) != Success )
						break;

					slot->exec( (char *) data );
					XFree( data );
				}
				break;
			case ButtonRelease:
				switch( event.xbutton.button )
				{
					case 1:
						{
							if( getMarkType() == Render::CrossMark )
							{
								int l = getMarkLeft();
								int t = getMarkTop();
								int x = event.xbutton.x;
								int y = event.xbutton.y;

								x -= slot->getLeft();
								y -= slot->getTop();

								if( l < 0 )
									l = slot->getWidth()-6+l;
								if( t < 0 )
									t = slot->getHeight()-6+t;

								if( x > l && x < l+8 &&
									y > t && y < t+8 )
								{
									slot->closeApplication();
									break;
								}
							}

							if( settings.getLeftClickAction() != 
								Settings::DoNothing &&
								slot->hasFocus() )
							{
								switch( settings.getLeftClickAction() )
								{
									case Settings::IconifyApplication:
										slot->iconifyApplication();
										continue;
									case Settings::LowerApplication:
										slot->lowerApplication();
										continue;
									case Settings::CloseApplication:
										slot->closeApplication();
										continue;
								}
							}

							slot->exec();
						}
						break;
					case 2:
						{
							Window owner;

							if( (owner = XGetSelectionOwner( display, 
								XA_PRIMARY )) == None )
								break;

							Atom property = XInternAtom( display, 
								"BBDOCK_SELECTION", False );

							XConvertSelection( display, XA_PRIMARY, XA_STRING,
								property, event.xany.window, CurrentTime );
						}
						break;
					case 3:
						switch( settings.getRightClickAction() )
						{
							case Settings::IconifyApplication:
								slot->iconifyApplication();
								break;
							case Settings::LowerApplication:
								slot->lowerApplication();
								break;
							case Settings::CloseApplication:
								slot->closeApplication();
								break;
						}
						break;
				}
				break;
		}
	}
}

/**
 * Send a message to a running instance of bbdock
 *
 * @param message - the message to send
 */
void Dock::send( const char *message, const char *data, int len )
{
	if( !message || !*message ||
		!data || !*data )
		return;

	Display *display;
	int screen;
	Window root;

	if(	!(display = XOpenDisplay( 0 )) )
		throw "Unable to open display";

	screen = DefaultScreen( display );

	if( !(root = RootWindow( display, screen )) )
		throw "Can not access root window";

	Window dummy;
	Window *childs;
	unsigned int n;

 	XQueryTree( display, root, &dummy, &dummy, &childs, &n );

	Atom msg = XInternAtom( display, message, false );

	for( ; n--; )
	{
		XClassHint xch;

		if( !XGetClassHint( display, childs[n], &xch ) )
			continue;

		if( !strcmp( xch.res_name, "bbdock" ) &&
			!strcmp( xch.res_class, "bbdock" ) )
		{
			XChangeProperty( display, childs[n], msg, msg, 8, 
				PropModeReplace, (unsigned char *) data, len );
		
			XEvent event;

			event.xclient.type = ClientMessage;
			event.xclient.serial = 0;
			event.xclient.send_event = True;
			event.xclient.message_type = msg;
			event.xclient.window = childs[n];
			event.xclient.format = 32;
			event.xclient.data.l[0] = 0;
			event.xclient.data.l[1] = 0;
			event.xclient.data.l[2] = 0;
			event.xclient.data.l[3] = 0;
			event.xclient.data.l[4] = 0;

			XSendEvent( display, childs[n], True, NoEventMask, &event );

			// exit loop but not without freeing resources below
			n = 0;
		}

		XFree( xch.res_name );
		XFree( xch.res_class );
	}

	if( childs )
		XFree( childs );

	XCloseDisplay( display );
}

/**
 * Change icon of a command (remotely)
 *
 * @param cmd - command of icon to change
 * @param icon - new icon
 */
void Dock::changeIcon( const char *cmd, const char *icon )
{
	if( !cmd || !*cmd ||
		!icon || !*icon )
		return;

	string data = cmd;

	data += ":";
	data += icon;

	Dock::send( ChangeIconMessage, data.c_str(), data.length() );
}

/**
 * Execute icon whose command is cmd (remotely)
 *
 * @param cmd - command of icon to change
 */
void Dock::executeIcon( const char *cmd )
{
	if( !cmd || !*cmd )
		return;

	Dock::send( ExecuteIconMessage, cmd, strlen( cmd ) );
}

/**
 * Process a client message
 *
 * @param e - event
 */
void Dock::client( XEvent *e )
{
	char *msg;

	if( !(msg = XGetAtomName( display, 
		e->xclient.message_type )) )
		return;

	Atom returnedtype;
	int format;
	unsigned long n;
	unsigned long bytesafter;
	char *prop;

	if( XGetWindowProperty( display, e->xany.window, e->xclient.message_type, 
		0, 1024, False, e->xclient.message_type, &returnedtype, &format, &n, 
		&bytesafter, (unsigned char **) &prop ) != Success )
		return;

	if( !strcmp( msg, ChangeIconMessage ) )
	{
		char *cmd;
		char *icon;

		if( (cmd = strtok( prop, ":" )) &&
			(icon = strtok( 0, "\n" )) )
		{
			Slot *slot;

			for( slot = first;
				slot;
				slot = slot->getNext() )
				if( !strcasecmp( slot->getCommand(), cmd ) && 
					strcasecmp( slot->getIcon(), icon ) )
				{
					slot->setIcon( icon );
					break;
				}
		}
	}
	else if( !strcmp( msg, ExecuteIconMessage ) )
	{
		Slot *slot;

		for( slot = first;
			slot;
			slot = slot->getNext() )
			if( !strcasecmp( slot->getCommand(), prop ) )
			{
				slot->exec();
				break;
			}
	}

	XFree( prop );
	XDeleteProperty( display, e->xany.window, e->xclient.message_type );

	XFree( msg );
}


src/Icon.cc
/*
 * Copyright (c) 2006 Markus Fisch
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. Neither the name of Markus Fisch nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS ``AS IS''
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */
#include <string>
#include <vector>

#include <stdlib.h>

#include "Icon.hh"

using namespace std;
using namespace bbdock;

/**
 * Initialize icon
 *
 * @param i - filename of image file
 * @param c - command to execute
 * @param w - window title if application should run exclusively (optional)
 * @param ms - idle time in miliseconds (optional)
 */
Icon::Icon( string i, string c, string t, unsigned int ms ) :
	image( expand( i ) ),
	command( expand( c ) ),
	idletime( ms )
{
	setTitle( t );
}

/**
 * Shell expansion
 *
 * @param path - path to expand
 */
string Icon::expand( string path ) const
{
	if( path[0] == '~' )
		path = getenv( "HOME" )+path.substr( 1 );

	return path;
}

/**
 * Tokenize string into vector
 *
 * @param str - string to split
 * @param tokens - vector that receives the tokens
 * @param delimiter - delimiter characters
 */
void Icon::tokenize( const string str, vector<string> &tokens,
	const string& delimiters ) const
{
    string::size_type last = str.find_first_not_of( delimiters, 0 );
    string::size_type next = str.find_first_of( delimiters, last );

    while( next != string::npos || last != string::npos )
    {
		tokens.push_back( str.substr( last, next-last ) );

		last = str.find_first_not_of( delimiters, next );
		next = str.find_first_of( delimiters, last );
    }
}


src/main.cc
/*
 * Copyright (c) 2006 Markus Fisch
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. Neither the name of Markus Fisch nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS ``AS IS''
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */
#include <sys/types.h>
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <unistd.h>
#include <iostream>
#include <fstream>

#include "Dock.hh"
#include "Render.hh"

using namespace std;
using namespace bbdock;

/**
 * Add an Icon record to a Dock
 *
 * @param dock - Dock object
 * @param record - an icon record
 * @param delimiter - field delimiter
 */
bool add( Dock &dock, char *record, const char delimiter )
{
	// cut off comments, don't use strtok because it will skip leading
	// delimiters
	{
		char *comment;

		if( (comment = strchr( record, '#' )) )
			*comment = 0;
	}

	enum
	{
		Image,
		Command,
		Title,
		Idle
	};

	char *fields[Idle+1];

	bzero( fields, sizeof( fields ) );
	fields[Image] = record;

	for( int n = Image; n <= Idle; )
	{
		char *stop;

		if( !(stop = strchr( fields[n], delimiter )) )
			break;

		*(stop++) = 0;
		fields[++n] = stop;
	}

	if( !fields[Image] || 
		!fields[Command] )
		return false;

	unsigned int idletime = Icon::Slow;

	if( fields[Idle] )
		if( !strcasecmp( fields[Idle], "fast" ) )
			idletime = Icon::Fast;
		else if( !strcasecmp( fields[Idle], "fastest" ) )
			idletime = Icon::Fastest;
		else if( !strcasecmp( fields[Idle], "slow" ) )
			idletime = Icon::Slow;
		else if( !strcasecmp( fields[Idle], "lame" ) )
			idletime = Icon::Lame;
		else 
		{
			unsigned int t = (unsigned int) 
				atoi( fields[Idle] );

			if( t > 0 )
				idletime = t;
		}

	return dock.add( new Icon( fields[Image], fields[Command], 
		(fields[Title] ? fields[Title] : ""), idletime ) );
}

/**
 * Load icons from file
 *
 * @param binary - name of the currently running binary
 * @param dock - Dock object
 */
void loadRC( char *binary, Dock &dock )
{
	string rcfilename = binary;
	string::size_type p;

	if( (p = rcfilename.find_last_of( '/' )) )
		rcfilename = rcfilename.substr( ++p );

	rcfilename = "/."+rcfilename+"rc";
	rcfilename = getenv( "HOME" )+rcfilename;

	ifstream fin( rcfilename.c_str(), ios::in );

	if( !fin )
		throw "Unable to open resource configuraton file for reading";

	for( string buf; getline( fin, buf ); )
		add( dock, (char *) buf.c_str(), ':' );
}

/**
 * Entry point
 */
int main( int argc, char **argv )
{
	Dock::Settings settings;
	char *binary;

	// process command line arguments  
	for( binary = *argv; --argc; )
		if( **(++argv) != '-' )
			break;
		else
			switch( *((*argv)+1) )
			{
				default:
					cerr << "Unknown argument \"" << *argv << "\" !" << endl 
						<< endl;
					// fall through
				case '?':
				case 'h':
					cout << "usage: " << binary << 
" [-hvdmprlcix] IMAGEFILE:COMMAND[:WINDOWTITLE;...[:IDLE]]...\n\
  -h                    print this help\n\
  -v                    print version\n\
  -d WIDTHxHEIGHT       outer dimensions of dock buttons\n\
  -m TYPE               define look of mark indicating a running instance,\n\
                        TYPE may be \"play\", \"dots\", \"corner\" or\n\
                        \"cross\" (default)\n\
  -p LEFT/TOP           left/top (or if negative right/bottom) padding of\n\
                        the mark in icon image\n\
  -r ACTION             determines what action is performed when a running\n\
                        icon gets right-clicked, possible ACTIONs are:\n\
                        \"nothing\" - do nothing\n\
                        \"iconfiy\" - iconify instance (default)\n\
                        \"lower\" - move instance to background\n\
                        \"close\" - shut down instance\n\
  -l ACTION             determines what action is performed when a icon\n\
                        is clicked again when the corresponding window is\n\
                        already activated, use the same ACTIONs like before,\n\
                        \"nothing\" is default\n\
  -c                    match WINDOWTITLE case-sensitive (recommended)\n\
  -i COMMAND:IMAGEFILE  remotely exchange icon of this command\n\
  -x COMMAND            remotely execute icon with this command\n\
\n\
IMAGEFILE   - should be path and filename of some PNG icon\n\
COMMAND     - a script or binary to execute\n\
WINDOWTITLE - is a semicolon-seperated list of case-insensitive window-titles\n\
              of corresponding application-windows. Those strings may contain\n\
              wildcard characters (* and/or ?) to exclusively  identify a\n\
              window. By providing this list you make the icon exclusive to\n\
              one instance of course. Clicking on already launched icons will\n\
              raise the corresponding window instead of invoking a new\n\
              instance.\n\
IDLE        - idle time after triggering one icon in miliseconds, instead\n\
              of using numbers you may also use the terms \"lame\" (~ 10 s),\n\
              \"slow\" (~ 5 s), \"fast\" (~ 500 ms) or \"fastest\" (~ 250 ms)"
						<< endl;
					return 0;
				case 'v':
					cout << "bbdock, version 0.2.9 <mf@markusfisch.de>" << endl;
					return 0;
				case 'd':
					{
						char *w;
						char *h;

						if( --argc &&
							(w = strtok( *(++argv), "x" )) &&
							(h = strtok( 0, "" )) )
						{
							settings.setSlotWidth( atoi( w ) );
							settings.setSlotHeight( atoi( h ) );
						}
						else
							cerr << "Missing or invalid dimensions !" << 
								endl;
					}
					break;
				case 'm':
					if( !--argc )
						cerr << "Missing mark type !" << endl;
					else
					{
						char *type = *(++argv);

						if( !(strcasecmp( type, "play" )) )
							settings.setMarkType( Render::PlayMark );
						else if( !(strcasecmp( type, "dots" )) )
						{
							settings.setMarkType( Render::DotsMark );
							settings.setMarkLeft( -1 );
							settings.setMarkTop( -1 );
						}
						else if( !(strcasecmp( type, "corner" )) )
						{
							settings.setMarkType( Render::CornerMark );
							settings.setMarkLeft( -1 );
							settings.setMarkTop( -1 );
						}
						else if( !(strcasecmp( type, "cross" )) )
						{
							settings.setMarkType( Render::CrossMark );
							settings.setMarkLeft( -1 );
							settings.setMarkTop( 0 );
						}
						else
							cerr << "Unknown type '" << type << "' !" << endl;
					}
					break;
				case 'p':
					{
						char *l;
						char *t;

						if( --argc &&
							(l = strtok( *(++argv), "/" )) &&
							(t = strtok( 0, "" )) )
						{
							settings.setMarkLeft( atoi( l ) );
							settings.setMarkTop( atoi( t ) );
						}
						else
							cerr << "Missing or invalid padding !" << 
								endl;
					}
					break;
				case 'r':
					if( !--argc )
						cerr << "Missing right-click action !" << endl;
					else
					{
						char *action = *(++argv);

						if( !strcasecmp( action, "nothing" ) )
							settings.setRightClickAction( 
								Dock::Settings::DoNothing );
						else if( !strcasecmp( action, "iconify" ) )
							settings.setRightClickAction( 
								Dock::Settings::IconifyApplication );
						else if( !strcasecmp( action, "lower" ) )
							settings.setRightClickAction( 
								Dock::Settings::LowerApplication );
						else if( !strcasecmp( action, "close" ) )
							settings.setRightClickAction( 
								Dock::Settings::CloseApplication );
						else
							cerr << "Unknown action '" << action << "' !" << 
								endl;
					}
					break;
				case 'l':
					if( !--argc )
						cerr << "Missing left-click action !" << endl;
					else
					{
						char *action = *(++argv);

						if( !strcasecmp( action, "nothing" ) )
							settings.setLeftClickAction( 
								Dock::Settings::DoNothing );
						else if( !strcasecmp( action, "iconify" ) )
							settings.setLeftClickAction( 
								Dock::Settings::IconifyApplication );
						else if( !strcasecmp( action, "lower" ) )
							settings.setLeftClickAction( 
								Dock::Settings::LowerApplication );
						else if( !strcasecmp( action, "close" ) )
							settings.setLeftClickAction( 
								Dock::Settings::CloseApplication );
						else
							cerr << "Unknown action '" << action << "' !" << 
								endl;
					}
					break;
				case 'c':
					settings.setCaseSensitive( true );
					break;
				case 'i':
					{
						char *icon;
						char *cmd;

						if( --argc &&
							(cmd = strtok( *(++argv), ":" )) &&
							(icon = strtok( 0, "" )) )
						{
							Dock::changeIcon( cmd, icon );
							return 0;
						}
						else
							cerr << "Missing or invalid parameter !" << 
								endl;
					}
					break;
				case 'x':
					{
						if( --argc )
						{
							Dock::executeIcon( *(++argv) );
							return 0;
						}
						else
							cerr << "Missing or invalid parameter !" << 
								endl;
					}
					break;
			}

	// detach from shell
	switch( fork() )
	{
		case -1:
			cerr << "Can not fork !" << endl;
			return -1;
		case 0:
			break;
		default:
			return 0;
	}

	// bring up the icons
	try
	{
		Dock dock( settings );

		if( !argc )
		{
			// try to start up by configuration file
			loadRC( binary, dock );
		}
		else
		{
			// try to start up by command line
			for( ; argc--; argv++ )
				if( !add( dock, *argv, ':' ) )
					throw "Invalid argument";
		}

		dock.run();
	}
	catch( const char *e )
	{
		cerr << e << endl;
	}
	catch( ... )
	{
		cerr << "Unknown error" << endl;
	}

	return 0;
}
